internal/stream/manager.go

```diff
@@ -0,0 +1,99 @@
+package stream
+
+import (
+	"context"
+
+	"github.com/razorpay/metro/internal/subscriber"
+	"github.com/razorpay/metro/internal/subscription"
+	"github.com/razorpay/metro/pkg/httpclient"
+	"github.com/razorpay/metro/pkg/logger"
+)
+
+// PushStreamManager manages push stream
+type PushStreamManager struct {
+	ctx        context.Context
+	cancelFunc func()
+	doneCh     chan struct{}
+	config     *httpclient.Config
+	ps         *PushStream
+}
+
+// NewPushStreamManager return a push stream manager obj which is used to manage push stream
+func NewPushStreamManager(ctx context.Context, nodeID string, subName string, subscriptionCore subscription.ICore, subscriberCore subscriber.ICore, config *httpclient.Config) (*PushStreamManager, error) {
+	ps, err := newPushStream(ctx, nodeID, subName, subscriptionCore, subscriberCore, config)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancelFunc := context.WithCancel(ctx)
+	return &PushStreamManager{
+		ctx:        ctx,
+		cancelFunc: cancelFunc,
+		ps:         ps,
+		doneCh:     make(chan struct{}),
+		config:     config,
+	}, nil
+}
+
+// Run starts the push stream manager that is used to manage underlying stream
+func (psm *PushStreamManager) Run() {
+	defer close(psm.doneCh)
+
+	logger.Ctx(psm.ctx).Infow("push stream manager: started running stream manager", "subscription", psm.ps.subscription.Name)
+	psm.startPushStream()
+
+	go func() {
+		for {
+			select {
+			case <-psm.ctx.Done():
+				if err := psm.ps.Stop(); err != nil {
+					logger.Ctx(psm.ctx).Infow("push stream manager: error stopping stream", "subscription", psm.ps.subscription.Name, "error", err.Error())
+				}
+				return
+			case err := <-psm.ps.GetErrorChannel():
+				logger.Ctx(psm.ctx).Infow("push stream manager: restarting stream handler", "subscription", psm.ps.subscription.Name, "error", err.Error())
+				psm.restartPushStream()
+			}
+		}
+	}()
+}
+
+// Stop stops the stream manager along with the underlying stream
+func (psm *PushStreamManager) Stop() {
+	logger.Ctx(psm.ctx).Infow("push stream manager: stop invoked", "subscription", psm.ps.subscription.Name)
+	psm.cancelFunc()
+	<-psm.doneCh
```

Comment:
this is the usual pattern which is followed across subscriber and stream modules. Not sure what else can be done here to avoid this. Don't want to do this asyncronously